fuzz-test-suite/dateparserfuzzer.cpp

```diff
@@ -0,0 +1,20 @@
+#include <ql/time/asx.hpp>
+#include <ql/time/date.hpp>
+#include <ql/time/ecb.hpp>
+#include <ql/time/imm.hpp>
+#include <ql/time/timeunit.hpp>
+#include <ql/utilities/dataparsers.hpp>
+#include <string>
+#include <fuzzer/FuzzedDataProvider.h>
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size) {
+    FuzzedDataProvider fdp(data, size);
+    constexpr int kMaxString = 100;
+    try {
+        (void)QuantLib::DateParser::parseFormatted(fdp.ConsumeRandomLengthString(kMaxString), "%Y-%m-%d");
+        (void)QuantLib::DateParser::parseISO(fdp.ConsumeRandomLengthString(kMaxString));
```

Comment:
Is this a random string of length up to 100?  If so, it's unlikely that it will ever be a date in ISO format.  Is the purpose of this test only to throw random data at the library and check that it doesn't crash or overwrite memory it doesn't own?

Reply:
That's mostly correct. Although I would add that the fuzzer will likely "learn" what the iso date format is. So it'll start off passing in seemingly random data, and then it'll measure the code coverage. The inputs that acheive the highest code-coverage will then be mutated to create new inputs. So it's likely that after a period of time the fuzzer will "learn" (through genetic programming), what the iso date format is, as that will produce the highest code coverage. We can help this process along and make it a bit quicker by providing a seed-corpus or a dictionary. But I haven't included either in the PR as I didn't want to complicate things to early on.

Reply:
I see, thanks.  Right now we're not training it, though, are we?  I'm guessing we should add something in the catch clause that tells the fuzzer the result of the attempt, right?

In your experience, is this something that mostly works for simple classes like this one, or does it scale to cases with a few dozens of inputs?  Real-world cases for this library would look at least like https://github.com/lballabio/QuantLib/blob/master/Examples/MulticurveBootstrapping/MulticurveBootstrapping.cpp which needs a whole lot of quotes and swap data.